### PR TITLE
Fixed an isolation level bug that could lead to unexpected isolation …

### PR DESCRIPTION
**Pull request by chrischu**
…levels when reusing db connections from the connection pool.

Should fix issue #17.

### CODE REVIEW

**Comment by alkampfergit** [comment]
@chrischu I'm sorry for the delay, I've tried to run tests against your https://github.com/chrischu/NEventStore.Persistence.SQL/tree/fix-isolation-level-bug branch, but actually I have two of your tests about transaction that are failing. (and they fails also for the pull-request)

What I do not understand is: at the very beginning of the test I found this comment

```
// This test proves that the bug still exists.
```

And this makes me believe that the fix is still not completed. To be accepted a pull request should not fail tests. Can you please check?

Here are the tests that are failing.

image

Thanks.

**Comment by chrischu** [comment]
What version of SQL Server are you using to run the tests?

On my machine (SQL Server 2014 SP1CU4) the tests run fine.

File: src/NEventStore.Persistence.MsSql.Tests/TransactionIsolationLevelTests.cs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+            Recorder.IsRecording = true;
+            Persistence.GetFrom();
+            Recorder.IsRecording = false;
+        }
+
+        [Fact]
+        public void should_run_command_in_non_default_isolation_level()
+        {
+            Recorder.StatementsWithIsolationLevels.Select(i => i.IsolationLevel)
+                .ShouldAllBeEquivalentTo(new[] {IsolationLevel.RepeatableRead});
```
Comment (fschmied) [inline src/NEventStore.Persistence.MsSql.Tests/TransactionIsolationLevelTests.cs]: @chrischu shouldn't this be {{ShouldAllBeEquivalentTo(new[] {IsolationLevel.ReadCommitted})}}?

**Comment by fschmied** [comment]
@alkampfergit The test called `when_reusing_a_connection_from_the_connection_pool_without_a_transaction_scope_and_not_circumventing_the_bug.should_run_command_in_non_default_isolation_level` is meant to prove that the fix is necessary, by reproducing the bug/unexpected behavior. I'm not sure if it's really required, especially since it depends on the SQL Server version being used.

**Comment by alkampfergit** [comment]
@fschmied If those two tests are not necessary we can simply delete them, it is not interesting to have a couple of test failing or depending from the Sql Server Version.

I'm using 

```
Microsoft SQL Server 2014 - 12.0.2269.0 (X64) 
	Jun 10 2015 03:35:45 
	Copyright (c) Microsoft Corporation
	Developer Edition (64-bit) on Windows NT 6.3 <X64> (Build 10586: ) (Hypervisor)
```

Following are the error in the build. The problem is that, I'm seeing those errors even running tests directly from Visual Studio. 

```
NEventStore.Persistence.MsSql.Tests.dll: NEventStore.Persistence.AcceptanceTests (2)
  when_reusing_a_connection_from_the_connection_pool_without_a_transaction_scope.should_run_command_in_non_default_isolation_level	
 
First failure:  	18/merge	#6.1.0-PullRequest.18+6	No changes	04 Apr 16 10:27
Expected item[0] to be 65536, but found 4096.

With configuration:
- Use declared types and members
- Compare enums by value
- Match member by name (or throw)
- Be strict about the order of items in byte arrays

   at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message) in z:\Builds\work\b7ceef20fdfcf0ee\Shared\Execution\LateBoundTestFramework.cs:line 25
   at FluentAssertions.Execution.CollectingAssertionStrategy.ThrowIfAny(IDictionary`2 context) in z:\Builds\work\b7ceef20fdfcf0ee\FluentAssertions.Core\Execution\CollectingAssertionStrategy.cs:line 50
   at FluentAssertions.Execution.AssertionScope.Dispose() in z:\Builds\work\b7ceef20fdfcf0ee\FluentAssertions.Core\Execution\AssertionScope.cs:line 252
   at FluentAssertions.Equivalency.EquivalencyValidator.AssertEquality(EquivalencyValidationContext context) in z:\Builds\work\b7ceef20fdfcf0ee\FluentAssertions.Core\Equivalency\EquivalencyValidator.cs:line 36
   at FluentAssertions.AssertionExtensions.ShouldBeEquivalentTo[T](T subject, Object expectation, String because, Object[] reasonArgs) in z:\Builds\work\b7ceef20fdfcf0ee\Shared\AssertionExtensions.cs:line 503
   at FluentAssertions.AssertionExtensions.ShouldAllBeEquivalentTo[T](IEnumerable`1 subject, IEnumerable expectation, String because, Object[] reasonArgs) in z:\Builds\work\b7ceef20fdfcf0ee\Shared\AssertionExtensions.cs:line 567
« Hide stacktrace
  when_reusing_a_connection_from_the_connection_pool_without_a_transact…ircumventing_the_bug.should_run_command_in_non_default_isolation_level	
 
First failure:  	18/merge	#6.1.0-PullRequest.18+6	No changes	04 Apr 16 10:27
Expected item[0] to be 65536, but found 4096.

With configuration:
- Use declared types and members
- Compare enums by value
- Match member by name (or throw)
- Be strict about the order of items in byte arrays

   at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message) in z:\Builds\work\b7ceef20fdfcf0ee\Shared\Execution\LateBoundTestFramework.cs:line 25
   at FluentAssertions.Execution.CollectingAssertionStrategy.ThrowIfAny(IDictionary`2 context) in z:\Builds\work\b7ceef20fdfcf0ee\FluentAssertions.Core\Execution\CollectingAssertionStrategy.cs:line 50
   at FluentAssertions.Execution.AssertionScope.Dispose() in z:\Builds\work\b7ceef20fdfcf0ee\FluentAssertions.Core\Execution\AssertionScope.cs:line 252
   at FluentAssertions.Equivalency.EquivalencyValidator.AssertEquality(EquivalencyValidationContext context) in z:\Builds\work\b7ceef20fdfcf0ee\FluentAssertions.Core\Equivalency\EquivalencyValidator.cs:line 36
   at FluentAssertions.AssertionExtensions.ShouldBeEquivalentTo[T](T subject, Object expectation, String because, Object[] reasonArgs) in z:\Builds\work\b7ceef20fdfcf0ee\Shared\AssertionExtensions.cs:line 503
   at FluentAssertions.AssertionExtensions.ShouldAllBeEquivalentTo[T](IEnumerable`1 subject, IEnumerable expectation, String because, Object[] reasonArgs) in z:\Builds\work\b7ceef20fdfcf0ee\Shared\AssertionExtensions.cs:line 567
```

**Comment by chrischu** [comment]
I just realized there was still a small mistake in one of the tests.

@alkampfergit Not BOTH tests are unnecessary, only one of them. One of them just verifies that the problem is still there in SQL Server (this is probably not necessary) and the other one verifies that the problem is actually fixed by my pull request.

I will remove the first one shortly.

**Comment by alkampfergit** [comment]
Thanks for the explanation. 

I've done other tests, and found that, if the test you removed is executed, then some subsequent tests are failing. I've investigated quickly the code, and I think there is a problem with 
[SqlPersistenceEngine::ExecuteQuery](https://github.com/NEventStore/NEventStore.Persistence.SQL/blob/accc7c627f4ad093e0dcd601043fc4eb511d95bd/src/NEventStore.Persistence.Sql/SqlPersistenceEngine.cs#L374)

In line 386 the transaction and connection object are not disposed after use. This could lead to problem, because noone closes transaction and connection after the query is executed.

If you look at the implementation of [SqlPersistenEngine::ExecuteCommand](https://github.com/NEventStore/NEventStore.Persistence.SQL/blob/accc7c627f4ad093e0dcd601043fc4eb511d95bd/src/NEventStore.Persistence.Sql/SqlPersistenceEngine.cs#L443) it correctly dispose everything.

The original implementation of OpenTransaction from MsDialect does nothing, thus it returns null, because it uses code from base class. Now if you always create a real Transaction to fix the bug of leaking isolation level, I suspect that some code could be broken because a connection transaction is still active. 

As an example, running tests with testdriven.net, with your original test suite, fails some test with this error (**ExecuteNonQuery requires the command to have a transaction when the connection assigned to the command is in a pending local transaction.  The Transaction property of the command has not been initialized.**)

```
Test 'Fixture NEventStore.Persistence.AcceptanceTests.when_specifying_a_hasher' failed: NEventStore.Persistence.StorageException : **ExecuteNonQuery requires the command to have a transaction when the connection assigned to the command is in a pending local transaction.  The Transaction property of the command has not been initialized**.
---- System.InvalidOperationException : ExecuteNonQuery requires the command to have a transaction when the connection assigned to the command is in a pending local transaction.  The Transaction property of the command has not been initialized.
	SqlPersistenceEngine.cs(475,0): at NEventStore.Persistence.Sql.SqlPersistenceEngine.ExecuteCommand[T](Func`3 command)
	SqlPersistenceEngine.cs(440,0): at NEventStore.Persistence.Sql.SqlPersistenceEngine.ExecuteCommand[T](Func`2 command)
	SqlPersistenceEngine.cs(268,0): at NEventStore.Persistence.Sql.SqlPersistenceEngine.Drop()
	Persistence\PipelineHooksAwarePersistanceDecorator.cs(118,0): at NEventStore.Persistence.PipelineHooksAwarePersistanceDecorator.Drop()
	WireupTests.cs(38,0): at NEventStore.Persistence.AcceptanceTests.when_specifying_a_hasher.Cleanup()
	BDD\SpecificationBase.cs(19,0): at NEventStore.Persistence.AcceptanceTests.BDD.SpecificationBase.OnFinish()
	BDD\SpecificationBaseRunner.cs(58,0): at NEventStore.Persistence.AcceptanceTests.BDD.SpecificationBaseRunner.ClassFinish()
	----- Inner Stack Trace -----
	at System.Data.SqlClient.SqlCommand.ValidateCommand(String method, Boolean async)
	at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, String methodName, Boolean sendToPipe, Int32 timeout, Boolean asyncWrite)
	at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
	SqlDialects\CommonDbStatement.cs(83,0): at NEventStore.Persistence.Sql.SqlDialects.CommonDbStatement.ExecuteNonQuery(String commandText)
	SqlPersistenceEngine.cs(268,0): at NEventStore.Persistence.Sql.SqlPersistenceEngine.<Drop>b__26_0(IDbStatement cmd)
	SqlPersistenceEngine.cs(440,0): at NEventStore.Persistence.Sql.SqlPersistenceEngine.<>c__DisplayClass39_0`1.<ExecuteCommand>b__0(IDbConnection _, IDbStatement statement)
	SqlPersistenceEngine.cs(455,0): at NEventStore.Persistence.Sql.SqlPersistenceEngine.ExecuteCommand[T](Func`3 command)
```

I'm not telling that your fix is wrong, but I believe that we should carefully review the ExecuteQuery to understand why it does not dispose automatically connection and transaction, before applying your fix.

Gian Maria.

**Comment by chrischu** [comment]
I just improved the test by making sure that everything is properly cleaned up after the test is run. 

Especially not enumerating the query I used to check if the isolation level is correct would have led to problems, since never enumerating it had the effect of not properly disposing the underlying commands/readers etc.

**Comment by alkampfergit** [comment]
@chrischu Closed. I've just one warn, the actual version of NEventStore has all SQL Persistence layer in the main NEventStore repository, and actually the package for Persistence.Sql is still not published, because it will be published along with Next NEventStore version.

This means that, probably, the same fix should be created on the actual master of NEventStore.

**Pull request merged**